### PR TITLE
add dependabot config script

### PR DESCRIPTION
**Pull request by DavidSpek**
Inspired by https://github.com/kubeflow/pipelines/issues/4682  I created a script that will create a config file for depandabot so that it knows what directories to scan. It will scan the repository for files named `*ockerfile*`, `package*.json`, `*requirements.txt` and `go.*`.  It is setup for dockerfiles and gomod at the moment. As I am not sure if the `go.mod` and `go.sum` in `./vendor` are relevant, I did not include searching for `go.*` files in the script. It is trivial to further customize what folders are selected if further customization is needed. It also parses the closest OWNERS file for a given directory and assigns the approvers and adds the reviewers to the PRs it creates.

This is a sibling PR to https://github.com/kubeflow/pipelines/pull/5015, https://github.com/kubeflow/kubeflow/pull/5542, https://github.com/kubeflow/kfserving/pull/1309 and https://github.com/kubeflow/arena/pull/403.

As it stands now, there are about 56 PRs that will be created with this configuration, so it might be advisable to have some form of plan to implement it in stages or be ready to quickly go through lots of the PRs. Another option is to create a target branch for all these PRs so they can be merged into that first rather than master.

For reference, the PRs that will be created can be found here: https://github.com/DavidSpek/testing/pulls

### CODE REVIEW

**Review by PatrickXYS — COMMENTED** [review]
Overall lgtm, some nit comments:

1. Have a make command to run create_dependabot.sh
2. Update README doc to have another section for dependabot functionality and description in this repo

File: hack/create_dependabot.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    append_updates('pip', string_path, assignees, reviewers)
+
+for go_path in get_go_paths():
+    string_path = str(go_path)
+    assignees = get_owners(go_path).get('approvers')
+    reviewers = get_owners(go_path).get('reviewers')
+    append_updates('gomod', string_path, assignees, reviewers)
+
+with open('.github/dependabot.yml', 'w') as outfile:
+    yaml.dump(dependabot, outfile, default_flow_style=False)
```
Comment (PatrickXYS) [inline hack/create_dependabot.py:96]: I think adding a main function will make this python file more organizable.

File: README.md [hunk 2]
```diff
@@ -1108,3 +1109,7 @@ is
   * TAG used for the images
 
 * Argo workflow should define the image paths and tag so that subsequent steps can use the newly built images
+
+## Creating dependabot config yaml for this repo
+
+In an effort to use the most current versions and mitigate vulnerable software dependencies and base images, a script was created to properly configure dependabot. The script scans the repository for directories containing files listing such dependencies, and matches the found folders to the relevant `OWNERS` files. It then goes on to generate the `.github/dependabot.yml` file which tells dependabot which directories it needs to scan and for what package ecosystems. When a dependency update is found, dependabot will create a pull request to update the dependency and assign the relevant owners. If changes are made to the repository that add new dependency listing files, the script will need to be run so that `.github/dependabot.yml` is updated to reflect these changes. To manually run the script, execute `make build-dependabot` from the root of this repository.
```
Comment (PatrickXYS) [inline README.md:1115]: Can we divide this paragraph into some smaller graphs? 

E.g, 

```
To use the most current versions and mitigate vulnerable software dependencies and base images, we configure dependabot for desired funtionality.

The way dependabot works as below:
1. We uses a script to scan the repository for directories containing files listing such dependencies, and matches the found folders to the relevant `OWNERS` files
2. Then it generate the `.github/dependabot.yml` file which tells dependabot which directories it needs to scan and for what package ecosystems.
3. When a dependency update is found, dependabot will create a pull request to update the dependency and assign the relevant owners.
...

And more details for dependabot can be found here (https://docs.github.com/en/github/managing-security-vulnerabilities/managing-vulnerabilities-in-your-projects-dependencies)

```

File: hack/create_dependabot.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
+import yaml
+import collections
+from pathlib import Path
+
+repo_path = Path(__file__).parents[1]
+
+print(repo_path.anchor)
```
Comment (PatrickXYS) [inline hack/create_dependabot.py:7]: Can we move this into main() func as well?  Is there a must reason we have to keep it outside?

Comment (DavidSpek) [reply hack/create_dependabot.py:7]: the print was for debugging so I will remove it, the repo_path I think can be added to main

File: README.md (hunk 2 above)
Comment (DavidSpek) [reply README.md]: I think the formatting is alright now. Not sure if there is a standard to adhere to.

**Comment by DavidSpek** [comment]
I am holding the PR to have some control over when it gets merged so that the optional test infra doesn't get overloaded if all the repo's were to merge this at the same time. 
/hold

**Comment by PatrickXYS** [comment]
Can we close the PR in favor of enabling Renovate?

**Comment by DavidSpek** [comment]
I’d say so. I’ll create a PR with an initial Renovate config based on the Pipelines one.

**Pull request closed without merging**